### ISSUE TITLE
Can't use MultiGPU in on channel mode

### ISSUE THREAD

**Issue by xiaomr** [Tencent/PocketFlow#64]
I prepared the horovod environment and tryed  to run on multi gpus.   whent I try
./scripts/run_local.sh nets/resnet_at_cifar10_run.py -n=4 -learner channel  according to the tutorials, the four gpus are occupied and training well.  But when I change to ./scripts/run_local.sh nets/resnet_at_cifar10_run.py -n=4 -learner dis-chn-pruned, only the first gpu is occupied and training, the others are empty.  Are multi gpus supported   on channel mode?

**Comment by jiaxiang-wu** [comment]
Hi, do you have horovod installed in the local Python environment? Try this (not in the docker):
``` Python
import horovod
```

**Comment by xiaomr** [comment]
@jiaxiang-wu horovod is installed OK  and the code is running in docker。when I using scripts/run_docker.sh , It only start docker and not run main.py,  so I have to use ./scripts/run_local.sh in the docker again

**Comment by jiaxiang-wu** [comment]
When you enter the docker environment, you should start the program with:
``` bash
$ bash main.sh
```
instead of
``` bash
$ ./run_local.sh ...
```
, since all extra parameters have been parsed by `scripts/run_docker.sh` (and will be used by `main.sh`), and you do not need pass them again to `scripts/run_local.sh`. The latter is only used when you want to run the program without docker.

**Comment by xiaomr** [comment]
@jiaxiang-wu  I try your method and but it will hang after print gpu info as following:

th 10390 MB memory) -> physical GPU (device: 3, name: GeForce GTX 1080 Ti, pci bus id: 0000:09:00.0, compute capability: 6.1)
2018-11-15 11:08:00.371735: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 2
2018-11-15 11:08:00.371809: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-11-15 11:08:00.371817: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977]      2
2018-11-15 11:08:00.371823: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 2:   N
2018-11-15 11:08:00.372240: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10390 MB memory) -> physical GPU (device: 2, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
2018-11-15 11:08:00.425454: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 1
2018-11-15 11:08:00.425532: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-11-15 11:08:00.425542: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977]      1
2018-11-15 11:08:00.425549: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 1:   N
2018-11-15 11:08:00.425994: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10390 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:05:00.0, compute capability: 6.1)
local-env:1133:1728 [0] NCCL INFO NCCL_IB_DISABLE set by environment to 1.
local-env:1133:1728 [0] NCCL INFO Using internal Network Socket
local-env:1133:1728 [0] NCCL INFO NET/Socket : 0 interfaces found

local-env:1133:1728 [0] transport/net_socket.cu:39 NCCL WARN NET/Socket : no interface found
local-env:1133:1728 [0] NCCL INFO transport/net_socket.cu:115 -> 3
local-env:1133:1728 [0] NCCL INFO bootstrap.cu:18 -> 3
local-env:1133:1728 [0] NCCL INFO bootstrap.cu:171 -> 3
local-env:1133:1728 [0] NCCL INFO bootstrap.cu:192 -> 3

**Comment by jiaxiang-wu** [comment]
Hi @xiaomr \
Can you execute horovod's example code in the docker? [Link to horovod's example code](https://github.com/uber/horovod/tree/master/examples)

**Comment by ginger0106** [comment]
> @jiaxiang-wu I try your method and but it will hang after print gpu info as following:
> 
> th 10390 MB memory) -> physical GPU (device: 3, name: GeForce GTX 1080 Ti, pci bus id: 0000:09:00.0, compute capability: 6.1)
> 2018-11-15 11:08:00.371735: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 2
> 2018-11-15 11:08:00.371809: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
> 2018-11-15 11:08:00.371817: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] 2
> 2018-11-15 11:08:00.371823: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 2: N
> 2018-11-15 11:08:00.372240: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10390 MB memory) -> physical GPU (device: 2, name: GeForce GTX 1080 Ti, pci bus id: 0000:08:00.0, compute capability: 6.1)
> 2018-11-15 11:08:00.425454: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 1
> 2018-11-15 11:08:00.425532: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
> 2018-11-15 11:08:00.425542: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977] 1
> 2018-11-15 11:08:00.425549: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 1: N
> 2018-11-15 11:08:00.425994: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10390 MB memory) -> physical GPU (device: 1, name: GeForce GTX 1080 Ti, pci bus id: 0000:05:00.0, compute capability: 6.1)
> local-env:1133:1728 [0] NCCL INFO NCCL_IB_DISABLE set by environment to 1.
> local-env:1133:1728 [0] NCCL INFO Using internal Network Socket
> local-env:1133:1728 [0] NCCL INFO NET/Socket : 0 interfaces found
> 
> local-env:1133:1728 [0] transport/net_socket.cu:39 NCCL WARN NET/Socket : no interface found
> local-env:1133:1728 [0] NCCL INFO transport/net_socket.cu:115 -> 3
> local-env:1133:1728 [0] NCCL INFO bootstrap.cu:18 -> 3
> local-env:1133:1728 [0] NCCL INFO bootstrap.cu:171 -> 3
> local-env:1133:1728 [0] NCCL INFO bootstrap.cu:192 -> 3

hi, I have the same problem, did you fix the problem? Thanks

**Comment by jiaxiang-wu** [comment]
@ginger0106 Does this problem occur only when you execute the PocketFlow code, or also occur when executing horovod's example code?

**Comment by ginger0106** [comment]
@jiaxiang-wu  I tried keras_mnist.py in horovod's example code, it worked as followed:

```
> root@local-env:/opt/ml/env# python keras_mnist.py 
> Using TensorFlow backend.
> 2018-11-20 04:01:22.351861: I tensorflow/core/platform/cpu_feature_guard.cc:141] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA
> 2018-11-20 04:01:25.123766: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1411] Found device 0 with properties: 
> name: GeForce GTX 1080 Ti major: 6 minor: 1 memoryClockRate(GHz): 1.582
> pciBusID: 0000:03:00.0
> totalMemory: 10.92GiB freeMemory: 10.76GiB
> 2018-11-20 04:01:25.123845: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1490] Adding visible gpu devices: 0
> 2018-11-20 04:01:25.684610: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971] Device interconnect StreamExecutor with strength 1 edge matrix:
> 2018-11-20 04:01:25.684698: I tensorflow/core/common_runtime/gpu/gpu_device.cc:977]      0 
> 2018-11-20 04:01:25.684717: I tensorflow/core/common_runtime/gpu/gpu_device.cc:990] 0:   N 
> 2018-11-20 04:01:25.685077: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1103] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10396 MB memory) -> physical GPU (device: 0, name: GeForce GTX 1080 Ti, pci bus id: 0000:03:00.0, compute capability: 6.1)
> Downloading data from https://s3.amazonaws.com/img-datasets/mnist.npz
> 11493376/11490434 [==============================] - 20s 2us/step
> x_train shape: (60000, 28, 28, 1)
> 60000 train samples
> 10000 test samples
> Train on 60000 samples, validate on 10000 samples
> Epoch 1/12
> 60000/60000 [==============================] - 6s 97us/step - loss: 0.2597 - acc: 0.9210 - val_loss: 
```

**Comment by ginger0106** [comment]
For the pocketflow,  I  also use 3 GPUs,  pull docker image uber/horovod:0.15.1-tf1.11.0-torch0.4.1-py3.5, change the index-url and  trusted-host in main.sh 
Then it shows that:
```
[local-env:00330] Read -1, expected 131072, errno = 1
[local-env:00331] Read -1, expected 131072, errno = 1
[local-env:00330] Read -1, expected 83968, errno = 1
[local-env:00331] Read -1, expected 83968, errno = 1
[local-env:00331] Read -1, expected 18816, errno = 1
[local-env:00330] Read -1, expected 18816, errno = 1
[local-env:00331] Read -1, expected 8192, errno = 1
[local-env:00330] Read -1, expected 8192, errno = 1
[local-env:00331] Read -1, expected 73728, errno = 1
[local-env:00330] Read -1, expected 73728, errno = 1
local-env:329:679 [0] NCCL INFO NCCL_IB_DISABLE set by environment to 1.
local-env:329:679 [0] NCCL INFO Using internal Network Socket
local-env:329:679 [0] NCCL INFO NET/Socket : 0 interfaces found

local-env:329:679 [0] transport/net_socket.cu:39 NCCL WARN NET/Socket : no interface found
local-env:329:679 [0] NCCL INFO transport/net_socket.cu:115 -> 3
local-env:329:679 [0] NCCL INFO bootstrap.cu:18 -> 3
local-env:329:679 [0] NCCL INFO bootstrap.cu:171 -> 3
local-env:329:679 [0] NCCL INFO bootstrap.cu:192 -> 3
```

**Comment by ginger0106** [comment]
hi @jiaxiang-wu, I tested `root@local-env:/opt/ml/env# mpirun -np 3 -H localhost:3 python keras_mnist.py` using 3GPUs, it could find 2 interfaces:
```
local-env:2445:2604 [0] misc/ibvwrap.cu:61 WARN Failed to open libibverbs.so[.1]
local-env:2445:2604 [0] INFO Using internal Network Socket
local-env:2445:2604 [0] INFO Using NCCL Low-latency algorithm for sizes below 16384
local-env:2445:2604 [0] INFO NET : Using interface eno1:10.1.1.244<0>
local-env:2445:2604 [0] INFO NET : Using interface eno2:192.168.0.2<0>
local-env:2445:2604 [0] INFO NET/Socket : 2 interfaces found
NCCL version 2.2.13+cuda9.0

local-env:2446:2598 [1] misc/ibvwrap.cu:61 WARN Failed to open libibverbs.so[.1]
local-env:2446:2598 [1] INFO Using internal Network Socket
local-env:2446:2598 [1] INFO Using NCCL Low-latency algorithm for sizes below 16384

local-env:2447:2601 [2] misc/ibvwrap.cu:61 WARN Failed to open libibverbs.so[.1]
local-env:2447:2601 [2] INFO Using internal Network Socket
local-env:2447:2601 [2] INFO Using NCCL Low-latency algorithm for sizes below 16384
local-env:2447:2601 [2] INFO comm 0x7fe834330190 rank 2 nranks 3
local-env:2447:2601 [2] INFO NET : Using interface eno1:10.1.1.244<0>
local-env:2447:2601 [2] INFO NET : Using interface eno2:192.168.0.2<0>
local-env:2447:2601 [2] INFO NET/Socket : 2 interfaces found
local-env:2446:2598 [1] INFO comm 0x7f86f43308d0 rank 1 nranks 3
local-env:2446:2598 [1] INFO NET : Using interface eno1:10.1.1.244<0>
local-env:2446:2598 [1] INFO NET : Using interface eno2:192.168.0.2<0>
local-env:2446:2598 [1] INFO NET/Socket : 2 interfaces found
local-env:2445:2604 [0] INFO comm 0x7f7adcb30dd0 rank 0 nranks 3
local-env:2445:2604 [0] INFO Using 256 threads
local-env:2445:2604 [0] INFO Min Comp Cap 6
local-env:2445:2604 [0] INFO NCCL_SINGLE_RING_THRESHOLD=131072
local-env:2445:2604 [0] INFO Ring 00 :    0   1   2
local-env:2447:2601 [2] INFO 2[2447] -> 0[2445] via direct shared memory
local-env:2446:2598 [1] INFO Ring 00 : 1[1] -> 2[2] via P2P/IPC
local-env:2445:2604 [0] INFO 0[2445] -> 1[2446] via direct shared memory
local-env:2445:2604 [0] INFO Launch mode Parallel

```

**Comment by xiaomr** [comment]
@jiaxiang-wu hi, I follow your instructions and it works in --learner dis-chn-pruned  and --learner weight-sparse in multi four gpu training . However, when I using --learner channel, all the other set is the same, It only training on the one of the four gpus (the other three gpu is free when -n=4).  I am wondered is there something wrong in channelpruning learner?

**Comment by xiaomr** [comment]
@ginger0106  You can try to change 
 options="-np ${nb_gpus} -H localhost:${nb_gpus} -bind-to none -map-by slot
      -x NCCL_DEBUG=INFO -x NCCL_SOCKET_IFNAME=eth1 -x NCCL_IB_DISABLE=1
      -x LD_LIBRARY_PATH --mca btl_tcp_if_include eth1"
to 
 options="-np ${nb_gpus} -H localhost:${nb_gpus}"  in main.sh  to fix the hang problem

**Comment by jiaxiang-wu** [comment]
@xiaomr 
You mentioned that `--learner dis-chn-pruned` and `--learner weight-sparse` works for multiple GPUs, but `--learner channel` does not. For `ChannelPrunedLearner`, does the program crash when multiple GPUs are used, or does it run normally, but only one GPU is used and others are idle?

**Comment by xiaomr** [comment]
@jiaxiang-wu  it run normally but only one GPU is used and others are idle,  I check the code and doubt whether it is because the channel selection and pruning procedure only use one gpu,  muti gpus will only be used in the last training pruned mode procedure.   Another question is how can I do with channel selection conflict  for adding identity path and shortcut path in resnet,  since in your implementation  the cout of convolution remain the same  because of the  substitution for 1*1 and k*k filters so the add conflict doesn't exist,  however , I hope to accelerate in GPU and manually remove the zero channels, but I can't find a good solution for the add conflict in resnet。

**Comment by jiaxiang-wu** [comment]
@xiaomr 
1. In `ChannelPrunedLearner`'s implementation, the channel selection is based on solving a LASSO problem on CPU (not GPU), and only one GPU is used for sampling input & output feature maps to construct this LASSO problem. After the channel selection process (this may take some time), the remaining training process should be running with multi-GPU support. We have included a GPU-based channel selection implementation, `ChannelPrunedGpuLearner`, in the pull request #84, and has been merged. You can try it out.
2. Directly editing the graph to remove those pruned channels during training is difficult. We do not have a good solution yet. Do you have any suggestions?

**Comment by xiaomr** [comment]
@jiaxiang-wu thank you very much for the detailed explaination!  It is really hard to editing graph during training so I have to edit graph after channel pruning  but before finetuning ,  so I can directly get the compressed model rather than the large but channel wise sparsity big model. As in your implementation, we can set some channels of some convolution and it's father convolution to zero after lasso, then finetune the pruned model, however , during the finetuning,  some zero channels may become non zero again,   have you encounter this problem?

**Comment by jiaxiang-wu** [comment]
@xiaomr 
Before fine-tuning, you need to compute mask tensors for convolutional kernels to mark channels that should be pruned. During fine-tuning, gradient tensors should be multiplied with these mask tensors before they are applied to update convolutional kernels.
